tst/unit/test_container_iterator.cpp

```diff
@@ -0,0 +1,139 @@
+
+//========================================================================================
+// Athena++ astrophysical MHD code
+// Copyright(C) 2014 James M. Stone <jmstone@princeton.edu> and other code contributors
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#include <iostream> // debug
+#include <memory>
+#include <string>
+#include <vector>
+#include <catch2/catch.hpp>
+
+#include "basic_types.hpp"
+#include "kokkos_abstraction.hpp"
+#include "interface/Metadata.hpp"
+#include "interface/Variable.hpp"
+#include "interface/Container.hpp"
+#include "interface/ContainerIterator.hpp"
+#include "parthenon_arrays.hpp"
+
+using parthenon::Real;
+using parthenon::Metadata;
+using parthenon::CellVariableVector;
+using parthenon::ParArrayND;
+using parthenon::Container;
+using parthenon::ContainerIterator;
+using parthenon::par_for;
+using parthenon::loop_pattern_mdrange_tag;
+using parthenon::DevSpace;
+
+TEST_CASE("Can pull variables from containers based on Metadata",
+          "[ContainerIterator]") {
+  GIVEN("A Container with a set of variables") {
+    Container<Real> rc;
+    Metadata m_in({Metadata::Independent});
+    Metadata m_out;
+    std::vector<int> scalar_block_size {16,16,16};
+    std::vector<int> vector_block_size {16,16,16,3};
+    // make some variables
+    rc.Add("v1",m_in,scalar_block_size);
+    rc.Add("v2",m_out,scalar_block_size);
+    rc.Add("v3",m_in,vector_block_size);
+    rc.Add("v4",m_out,vector_block_size);
+    rc.Add("v5",m_in,scalar_block_size);
+    rc.Add("v6",m_out,scalar_block_size);
+
+    WHEN("We initialize all variables to zero") {
+      // set them all to zero
+      const CellVariableVector<Real>& cv = rc.GetCellVariableVector();
+      for (int n=0; n<cv.size(); n++) {
```

Comment:
Thanks @jdolence ! This looks like a good starting point. The goal would be to move this layer of the looping into a kernel. Later on I'll add some performance tests  with some of the kernels that are likely to cause problems. Not in this PR though.